text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,165 @@
+
+```
+                               CONSUMER
+
+                   . . . . . . Settle
+PRODUCER           .              ^
+                   .              |
+Create             .           Process
+  |                v              ^
+  v        +--------------+       |            
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget), the settlement stage
+   does not exist.
+
+The semantic conventions described below define how to model those stages in
+traces, how to propagate context, and how to enrich traces with attributes.
+
+### Context propagation
+
+Two layers of context propagation need to be distinguished for messaging
+workflows:
+
+1. The _creation context layer_ allows correlating the producer with the
+   consumers of a message, regardless of intermediary instrumentation. The
+   creation context is created by the producer and must be propagated to the
+   consumers. It must not be altered by intermediaries.
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. The _transport context layer_ allows correlating the producer and the
+   consumer with an intermediary. It also allows to correlate multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries
+   that are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+A producer MUST attach a creation context to each message. The creation context
+MUST be attached in a way so that it is not changed by intermediaries. A
+producer MAY propagate a transport context to an intermediary.  An
+intermediary MAY propagate a transport context to a consumer.
+
+### Span structure, names, and attributes
+
+#### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
```

Comment:
I would also add (for documentation in this OTEP) that we can possibly have more than one "process" span per message.
correlating the consumer and producer based only on "process" requires a more complicated and error prone logic in tracing backends:
- backends need to aggregate spans to create a complete trace. this aggregation is heuristic by nature (how much time should we wait for spans to arrive?) and can be computationally expensive.
- backends need to traverse the trace and identify relevant "process" spans for each "receive". Even if "process" is a direct child of "receive", they still need to be filtered from possibly other transport and batch operation spans.
- In the presence of more than one "process" span per message, backends need to apply some filtering logic to drop duplicated occurrences of the same "create" context. While it is technically possible, it makes processing more complex

Reply:
I think we might add some more information about creating "Process" spans to the OTEP (although we won't require the creation of "Process" spans).  The proposed additions in this comment can then be considered there.

However, I think it requires some more discussions before we can do that.

Reply:
I would suggest extending the conventions by recommendations for creating process spans after semantic conventions are stabilized (this will be possible, as process spans will not be mandatory anyway). This would help us to gather some user experience and make more informed decisions later on. Let's discuss this in one of the next workgroup meetings.